File: algorithm/LeetCode/AddTwoNumbers/AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode * curr = new ListNode(0);
        ListNode * root = curr;
        ListNode * prev = curr;
        int currVal = 0;
        while(l1 != NULL && l2 != NULL){
            currVal = l1->val + l2->val;
            curr->val += currVal;
            curr->next = new ListNode(curr->val/10);
            curr->val%=10;
            prev = curr;
            curr = curr->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while(l1 != NULL){
            curr->val += l1->val;
            curr->next = new ListNode(curr->val/10);
            curr->val %= 10;
            prev = curr;
            curr = curr->next;
            l1 = l1->next;
        }
        while(l2 != NULL){
            curr->val += l2->val;
            curr->next = new ListNode(curr->val/10);
            curr->val %= 10;
            prev = curr;
            curr = curr->next;
            l2 = l2->next;
        }
        if(curr->val == 0){
            prev->next = NULL;
            delete curr;
        }
        return root;
    }
};

File: algorithm/LeetCode/AddTwoNumbers/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode tmp = new ListNode(0);
        ListNode head = tmp;
        int carry = 0;
        while(l1!=null || l2!=null || carry != 0){
            int val = ((l1 != null)?l1.val:0) + ((l2!=null)?l2.val:0) + carry;
            carry = carry/10 + val/10;
            val %= 10;
            tmp.next = new ListNode(val);
            tmp = tmp.next;
            l1 = l1!=null ? l1.next : l1;
            l2 = l2!=null ? l2.next : l2;
        }
        return head.next;
    }
}